File: DSA/LC/143_reorder_list.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if head is None:
            return None

        def rec(target: ListNode, curr: ListNode) -> ListNode | None:
            if curr is None:
                return target

            target = rec(target, curr.next)

            if target is None:
                return None

            if target.next is curr or target is curr:
                curr.next = None
                return None

            next_target = target.next
            target.next = curr
            curr.next = next_target
            target = next_target

            return target

        rec(head, head.next)

File: DSA/LC/143_reorder_list.cpp
struct ListNode
{
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
  void reorderList(ListNode *head)
  {
    ListNode *fastPtr = head->next;
    ListNode *slowPtr = head;
    while (fastPtr != nullptr && fastPtr->next != nullptr)
    {
      fastPtr = fastPtr->next->next;
      slowPtr = slowPtr->next;
    }

    ListNode *first = head;
    ListNode *second = slowPtr->next;
    slowPtr->next = nullptr;

    ListNode *curr = second;
    ListNode *prev = nullptr;
    ListNode *next = second;

    while (next)
    {
      next = curr->next;
      curr->next = prev;
      prev = curr;
      curr = next;
    }

    ListNode *firstNext = first;
    ListNode *secondNext = second;
    second = prev;
    while (second)
    {
      firstNext = first->next;
      secondNext = second->next;
      first->next = second;
      second->next = firstNext;
      first = firstNext;
      second = secondNext;
    }
  }
};